src/services/types.ts

```diff
@@ -0,0 +1,83 @@
+import { MaskLevel } from 'common/typedefs/Permission';
+import { Application, Group, Policy, User } from '../common/typedefs';
+
+type Entity = User | Group | Application | Policy;
+export type UpdateEntity<T> = ({ item }: { item: T }) => Promise<T>;
+export type DeleteEntity<T> = ({ item }: { item: T }) => Promise<string>;
+export type CreateEntity<T> = ({ item }: { item: Partial<T> }) => Promise<T>;
+
+export type AddToEntity<T> = ({
+  entity,
+  key,
+  value,
+}: {
+  entity: T;
+  key: string;
+  value: string;
+}) => Promise<T>;
+export type RemoveFromEntity<T> = ({
+  entity,
+  key,
+  value,
+}: {
+  entity: T;
+  key: string;
+  value: string;
+}) => Promise<T>;
+
+export type AddGroupToEntity<T> = ({ group, entity }: { group: Group; entity: T }) => Promise<T>;
+export type AddApplicationToEntity<T> = ({
+  application,
+  entity,
+}: {
+  application: Application;
+  entity: T;
+}) => Promise<T>;
+
+export type AddUserToEntity<T> = ({ user, entity }: { user: User; entity: T }) => Promise<T>;
+
+export type RemoveGroupFromEntity<T> = ({
+  group,
+  entity,
+}: {
+  group: Group;
+  entity: T;
+}) => Promise<T>;
+export type RemoveApplicationFromEntity<T> = ({
+  application,
+  entity,
+}: {
+  application: Application;
+  entity: T;
+}) => Promise<T>;
+export type RemoveUserFromEntity<T> = ({ user, entity }: { user: User; entity: T }) => Promise<T>;
+
+export interface GroupWithMask extends Group {
+  mask: MaskLevel;
+}
+
+export interface UserWithMask extends User {
+  mask: MaskLevel;
+}
+
+export interface PolicyWithMask extends Policy {
+  mask: MaskLevel;
+}
```

Comment:
why combine these instead of just passing the entity and the mask separately?
edit: i see your TODO comments. nevermind. 😨